### PR TITLE
feat: bind embedding models to vec stores and use vector store map

### PR DESCRIPTION
**Pull request by marcusschiesser**
- bind `embedModel` to vector stores
- use multiple vector stores and split them by the modality type that they should store
- add `vectorStores` to storage context to define vector store per modality

### CODE REVIEW

File: packages/core/src/indices/vectorStore/index.ts [hunk 1]
```diff
@@ -231,10 +212,12 @@ export class VectorStoreIndex extends BaseIndex<IndexDict> {
     }
 
     // use doc store strategy to avoid duplicates
+    // TODO: use doc strategy with multiple vector stores
```
Comment (marcusschiesser) [inline packages/core/src/indices/vectorStore/index.ts:215]:
```suggestion
```

File: packages/core/src/indices/vectorStore/index.ts [hunk 2]
```diff
@@ -248,20 +231,30 @@ export class VectorStoreIndex extends BaseIndex<IndexDict> {
     return await this.init(args);
   }
 
-  static async fromVectorStore(
-    vectorStore: VectorStore,
-    serviceContext?: ServiceContext,
-    imageVectorStore?: VectorStore,
-  ) {
-    if (!vectorStore.storesText) {
+  static async fromVectorStore({
```
Comment (marcusschiesser) [inline packages/core/src/indices/vectorStore/index.ts:234]: maybe better add `fromVectorStores`?

File: packages/core/src/indices/vectorStore/index.ts [hunk 3]
```diff
@@ -415,47 +402,35 @@ export class VectorIndexRetriever implements BaseRetriever {
...
     query: string,
+    vectorStore: VectorStore,
     preFilters?: MetadataFilters,
   ): Promise<NodeWithScore[]> {
     const options = {};
     const q = await this.buildVectorStoreQuery(
-      this.index.embedModel,
+      vectorStore.embedModel,
       query,
+      // TODO: use different topK for image and text
```
Comment (marcusschiesser) [inline packages/core/src/indices/vectorStore/index.ts:429]: TODO

File: packages/core/src/storage/StorageContext.ts [hunk 4]
```diff
@@ -53,20 +52,19 @@ export async function storageContextFromDefaults({
...
-          path.join(persistDir, DEFAULT_IMAGE_VECTOR_NAMESPACE),
+    if (!(ObjectType.TEXT in vectorStores)) {
+      vectorStores[ObjectType.TEXT] =
+        vectorStore ??
+        ((await SimpleVectorStore.fromPersistDir(
+          persistDir,
           fs,
-        )
-      : imageVectorStore;
+        )) as unknown as VectorStore);
```
Comment (marcusschiesser) [inline packages/core/src/storage/StorageContext.ts:61]: @himself65 this is not very nice. How can I make `SimpleVectorStore.fromPersistDir` return a `VectorStore` given the fact that it's build by the mixin (see source of SimpleVectorStore)?

File: examples/multimodal/load.ts [hunk 5]
```diff
@@ -25,12 +27,16 @@ async function generateDatasource() {
     const documents = await new SimpleDirectoryReader().loadData({
       directoryPath: path.join("multimodal", "data"),
     });
-    const storageContext = await storageContextFromDefaults({
-      persistDir: "storage",
-      storeImages: true,
+    const clipVectorStore = new SimpleVectorStore({
+      embedModel: new ClipEmbedding(),
```
Comment (marcusschiesser) [inline examples/multimodal/load.ts:31]: @himself65 `pnpm type-check` produces here:
```
Object literal may only specify known properties, and 'embedModel' does not exist in type '{ data?: SimpleVectorStoreData | undefined; }'.
```
doesn't make a lot of sense to me as `SimpleVectorStore` is produced by a mixin and the mixin's constructor allows any

File: examples/multimodal/rag.ts [hunk 6]
```diff
@@ -41,7 +41,7 @@ async function main() {
 
   const queryEngine = index.asQueryEngine({
     responseSynthesizer: new MultiModalResponseSynthesizer(),
-    retriever: index.asRetriever({ similarityTopK: 3, imageSimilarityTopK: 1 }),
+    retriever: index.asRetriever({ topK: { text: 3, image: 1 } }),
```
Comment (marcusschiesser) [inline examples/multimodal/rag.ts:44]: should probably better use uppercase here
```suggestion
    retriever: index.asRetriever({ topK: { TEXT: 3, IMAGE: 1 } }),
```

File: packages/core/src/ingestion/IngestionPipeline.ts [hunk 7]
```diff
@@ -176,3 +189,31 @@ export class IngestionPipeline {
...
+    newIds: string[],
+    nodes: BaseNode<Metadata>[],
+    vectorStore: VectorStore,
+  ) => Promise<void>,
+) {
+  const nodeMap = splitNodesByType(nodes);
+  for (const type in nodeMap) {
+    const nodes = nodeMap[type as ModalityType];
+    if (nodes) {
+      console.log(type);
```
Comment (marcusschiesser) [inline packages/core/src/ingestion/IngestionPipeline.ts:206]: remove

File: examples/multimodal/load.ts (hunk 5 above)
Comment (himself65) [reply examples/multimodal/load.ts:31]: I added some changes, I think type here is indeed wrong

File: .github/workflows/test.yml [hunk 8]
```diff
@@ -75,9 +75,6 @@ jobs:
         working-directory: ./examples
       - name: Run Type Check
         run: pnpm run type-check
-      - name: Run Circular Dependency Check
-        run: pnpm run circular-check
```
Comment (marcusschiesser) [inline .github/workflows/test.yml:79]: @himself65 adding the embedModel to the vector store base class added a lot of circular dependencies which seem to be hard to resolve.

I checked that `madge` is checking circular deps on a file basis, not on object level. Seems too strict to me - why did you add it?

Comment (himself65) [reply .github/workflows/test.yml:79]: No, check circular deps on file base makes sense. It will cause incorrect behavior for bundler like using Next.js
And there is no meaning to check on object level

Comment (himself65) [reply .github/workflows/test.yml:79]: Dont remove this

Comment (himself65) [reply .github/workflows/test.yml:79]: This is a very common bug, see https://stackoverflow.com/a/44727578
